File: Sources 4.5.2/Gizmox.WebGUI.Forms.CompanionKit/Controls/DataGridView/Programming/ProgramaticSortingPage.cs
#region Using

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using Gizmox.WebGUI.Common;
using Gizmox.WebGUI.Forms;
using System.Collections;

#endregion

namespace CompanionKit.Controls.DataGridView.Programming
{
    public partial class ProgramaticSortingPage : UserControl
    {
        private int _sortColumn = -1;
        private SortOrder _sortOrder = SortOrder.Ascending;

        private DataGridViewItemComparer _numericComparer = null;
        private DataGridViewItemComparer _stringComparer = null;

		public ProgramaticSortingPage()
        {
            InitializeComponent();
        }

        private void ProgramaticSortingPage_Load(object sender, EventArgs e)
        {
          
            // compare by number
			_numericComparer = new NumericComparer(0, SortOrder.Ascending);

			// compare by text
			_stringComparer = new StringComparer(1, SortOrder.Ascending, CaseInsensitiveComparer.Default);	
				
            // Fill up DataGridView
            for (int i = 1; i < 20; ++i)
            {
                this.mobjDataGridView.Rows.Add((20-i).ToString(), string.Format("User{0}", i), string.Format("user{0}@gmail.com", i),
                                         string.Format("8-800-236589{0}", i), "Franklin",
                                         string.Format("10{0} Murfreeboro Rd.", i), "USA", "Tennessee", "37064");
            }
        }

        private void dataGridView1_ColumnHeaderMouseClick(object sender, DataGridViewCellMouseEventArgs e)
        {
            // Call the sort method to manually sort.
            this.mobjDataGridView.Sort(GetComparatorForColumn(e.ColumnIndex));
        }

        /// <summary>
        /// Gets comparator for the specified column
        /// </summary>
        /// <param name="columnIndex"></param>
        /// <returns></returns>
        private DataGridViewItemComparer GetComparatorForColumn(int columnIndex)
        {
            // Select comparator according to column index
            DataGridViewItemComparer itemComparer = null;
            switch (columnIndex)
            {
                case 0:
                    itemComparer = _numericComparer;
                    break;
                default:
                    itemComparer = _stringComparer;
                    break;
            }

            itemComparer.Column = columnIndex;
            // Determine whether the column is the same as the last column clicked.
            if (columnIndex != _sortColumn)
            {
                // Set the sort column to the new column.
                _sortColumn = columnIndex;
                // Set the sort order to ascending by default.
                _sortOrder = SortOrder.Ascending;
            }
            else
            {
                // Determine what the last sort order was and change it.
				_sortOrder  = (_sortOrder == SortOrder.Ascending) ? SortOrder.Descending : SortOrder.Ascending;
				
            }
            itemComparer.Order = _sortOrder;

            return itemComparer;
        }

        // Implements comparator for the custom sorting of items by columns.
        public abstract class  DataGridViewItemComparer : IComparer
        {
            protected int column;
            protected SortOrder order;
            public DataGridViewItemComparer(int Column, SortOrder order)
            {
                this.column = Column;
                this.order = order;
            }

            public int Column
            {
                set
                {
                    column = value;
                }
            }

            public SortOrder Order
            {
                set
                {
                    order = value;
                }
            }

            public abstract int Compare(object x, object y);
        }

        /// <summary>
        /// Implements a class for comparison of DataGridView items
        /// according to a numeric data in the given column
        /// </summary>
        public class NumericComparer : DataGridViewItemComparer
        {
            public NumericComparer(int Column, SortOrder order)
                :base(Column, order)
            {
            }

            public override int Compare(object x, object y)
            {
                DataGridViewRow dataGridViewRow1 = (DataGridViewRow)x;
                DataGridViewRow dataGridViewRow2 = (DataGridViewRow)y;
                int returnVal =
                    Int32.Parse(dataGridViewRow1.Cells[column].Value.ToString()) -
                    Int32.Parse(dataGridViewRow2.Cells[column].Value.ToString());

                return order == SortOrder.Ascending ? returnVal : returnVal * -1;
            }
        }

        /// <summary>
        /// Implements a class for comparison of DataGridView items
        /// according to a literal data in the given column
        /// by the given IComparer object
        /// </summary>		
        public class StringComparer : DataGridViewItemComparer
        {
            IComparer mobjComparer = null;

            public StringComparer(int Column, SortOrder order, IComparer StringComparer)
                :base(Column, order)
            {
                mobjComparer = StringComparer;
            }

            public override int Compare(object x, object y)
            {
                DataGridViewRow dataGridViewRow1 = (DataGridViewRow)x;
                DataGridViewRow dataGridViewRow2 = (DataGridViewRow)y;
                int returnVal = mobjComparer.Compare(
                    dataGridViewRow1.Cells[column].Value.ToString(),
                    dataGridViewRow2.Cells[column].Value.ToString());
                return order == SortOrder.Ascending ? returnVal : returnVal * -1;
            }
        }


    }
}

File: Sources 4.5.2/Gizmox.WebGUI.Forms.CompanionKit/Controls/DataGridView/Programming/ProgramaticSortingPage.vb
Namespace CompanionKit.Controls.DataGridView.Programming

	Public Class ProgramaticSortingPage

		Private _sortColumn As Integer = 0
		Private _sortOrder As SortOrder = SortOrder.Ascending
        Private _stringComparer As DataGridViewItemComparer = Nothing
        Private _numericComparer As DataGridViewItemComparer = Nothing


		Public Sub New()

			' This call is required by the Visual WebGui UserControl Designer.
			InitializeComponent()

		End Sub

        Private Sub ProgramaticSortingPage_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
            ' compare by number
            Me._numericComparer = New NumericComparer(0, SortOrder.Ascending)

            ' compare by text
            Me._stringComparer = New StringComparer(1, SortOrder.Ascending, CaseInsensitiveComparer.Default)
            ' Fill up DataGridView
            Dim i As Integer
            For i = 1 To 19
                Me.mobjDataGridView.Rows.Add((20 - i).ToString(), String.Format("User{0}", i), _
        String.Format("user{0}@gmail.com", i), _
        String.Format("8-800-236589{0}", i), "Franklin", _
        String.Format("10{0} Murfreeboro Rd.", i), "USA", "Tennessee", "37064")
            Next i
        End Sub

        Private Sub DataGridView1_ColumnHeaderMouseClick(ByVal sender As System.Object, ByVal e As Global.Gizmox.WebGUI.Forms.DataGridViewCellMouseEventArgs) Handles mobjDataGridView.ColumnHeaderMouseClick


            ' Call the sort method to manually sort.
            Me.mobjDataGridView.Sort(Me.GetComparatorForColumn(e.ColumnIndex))

        End Sub

        ''' <summary>
        ''' Gets comparator for the specified column
        ''' </summary>
        ''' <param name="columnIndex"></param>
        Private Function GetComparatorForColumn(ByVal columnIndex As Integer) As DataGridViewItemComparer
            Dim itemComparer As DataGridViewItemComparer = Nothing
            Select Case columnIndex
				Case 0
					itemComparer = Me._numericComparer
					Exit Select
                Case Else
                    itemComparer = Me._stringComparer
                    Exit Select
            End Select
            itemComparer.Column = columnIndex
            ' Determine whether the column is the same as the last column clicked.
            If (columnIndex <> Me._sortColumn) Then
                ' Set the sort column to the new column.
                Me._sortColumn = columnIndex
                ' Set the sort order to ascending by default.
                Me._sortOrder = SortOrder.Ascending
                ' Determine what the last sort order was and change it.
            ElseIf (Me._sortOrder = SortOrder.Ascending) Then
                Me._sortOrder = SortOrder.Descending
            Else
                Me._sortOrder = SortOrder.Ascending
            End If
            itemComparer.Order = Me._sortOrder
            Return itemComparer
        End Function


        ''' <summary>
        '''   Implements comparator for the custom sorting of items by columns.
        ''' </summary>
        Public MustInherit Class DataGridViewItemComparer
            Implements Global.System.Collections.IComparer

            ' Methods
            Public Sub New(ByVal Column As Integer, ByVal order As SortOrder)
                Me.Column = Column
                Me.Order = order
            End Sub

            ' Methods
            Public MustOverride Function [Compare](ByVal x As Object, ByVal y As Object) As Integer Implements System.Collections.IComparer.Compare


            ' Properties
            Public WriteOnly Property Column() As Integer
                Set(ByVal value As Integer)
                    Me._column = value
                End Set
            End Property

            Public WriteOnly Property Order() As SortOrder
                Set(ByVal value As SortOrder)
                    Me._order = value
                End Set
            End Property


            ' Fields
            Protected _column As Integer = 0
            Protected _order As SortOrder = SortOrder.Ascending
        End Class

        ''' <summary>
        ''' Implements a class for comparison of DataGridView items
        ''' according to a numeric data in the given column
        ''' </summary>
        ''' <remarks></remarks>
        Public Class NumericComparer
            Inherits DataGridViewItemComparer
            ' Methods
            Public Sub New(ByVal Column As Integer, ByVal order As SortOrder)
                MyBase.New(Column, order)
            End Sub

            Public Overrides Function [Compare](ByVal x As Object, ByVal y As Object) As Integer
                Dim dataGridViewRow1 As DataGridViewRow = DirectCast(x, DataGridViewRow)
                Dim dataGridViewRow2 As DataGridViewRow = DirectCast(y, DataGridViewRow)
                Dim returnVal As Integer = (Integer.Parse(dataGridViewRow1.Cells.Item(MyBase._column).Value.ToString) - _
                                            Integer.Parse(dataGridViewRow2.Cells.Item(MyBase._column).Value.ToString))
                Return IIf((MyBase._order = SortOrder.Ascending), returnVal, (returnVal * -1))
            End Function

        End Class

        ''' <summary>
        ''' Implements a class for comparison of DataGridView items
        ''' according to a literal data in the given column
        ''' by the given IComparer object
        ''' </summary>
        ''' <remarks></remarks>
        Public Class StringComparer
            Inherits DataGridViewItemComparer
            ' Methods
            Public Sub New(ByVal Column As Integer, ByVal order As SortOrder, ByVal StringComparer As IComparer)
                MyBase.New(Column, order)
                Me.mobjComparer = Nothing
                Me.mobjComparer = StringComparer
            End Sub

            Public Overrides Function [Compare](ByVal x As Object, ByVal y As Object) As Integer
                Dim dataGridViewRow1 As DataGridViewRow = DirectCast(x, DataGridViewRow)
                Dim dataGridViewRow2 As DataGridViewRow = DirectCast(y, DataGridViewRow)
                Dim returnVal As Integer = Me.mobjComparer.Compare(dataGridViewRow1.Cells.Item(MyBase._column).Value.ToString, _
                                                                   dataGridViewRow2.Cells.Item(MyBase._column).Value.ToString)
                Return IIf((MyBase._order = SortOrder.Ascending), returnVal, (returnVal * -1))
            End Function


            ' Fields
            Private mobjComparer As IComparer
        End Class

	End Class

End Namespace
